pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeSerDe.java

```diff
@@ -380,14 +402,76 @@ public static StarTreeInterf fromFile(File starTreeFile, ReadMode readMode)
       return new StarTreeV2(starTreeFile, readMode);
     } else {
       throw new RuntimeException("Unrecognized version for Star Tree " + starTreeVersion);
     }
   }
+
+  /**
+   * Utility method to convert star tree from v1 to v2 format:
+   * <p>- If star tree does not exist, or if actual version is the same as
+   *   expected version, then no action is taken. </p>
+   *
+   * <p>- If actual version is v1 and expected version is v2, then conversion from
+   *   v1 to v2 is performed. Both v1 and v2 formats are also backed up.</p>
+   *
+   * <p>- If actual version is v2 and expected version is v1, then v1 is restored from
+   *   backup version, if available, no-op otherwise. Note, there is no v2 to v1
+   *   conversion as of now.</p>
+   *
+   * @param indexDir
+   * @param starTreeVersionToLoad
+   * @throws IOException
+   * @throws ClassNotFoundException
+   */
+  public static void convertStarTreeFormatIfNeeded(File indexDir, StarTreeFormatVersion starTreeVersionToLoad)
+      throws IOException, ClassNotFoundException {
+    File starTreeFile = new File(indexDir, V1Constants.STAR_TREE_INDEX_FILE);
+
+    // If the star-tree file does not exist, this is not a star tree index, nothing to do here.
+    if (!starTreeFile.exists()) {
+      LOGGER.info("Skipping Star Tree format conversion, as no star tree file exists for {}",
+          starTreeFile.getAbsoluteFile());
+      return;
+    }
+
+    StarTreeFormatVersion actualVersion = getStarTreeVersion(starTreeFile);
+    if (actualVersion == StarTreeFormatVersion.V1 && starTreeVersionToLoad == StarTreeFormatVersion.V2) {
+      LOGGER.info("Converting Star Tree from V1 to V2 format for {}", starTreeFile.getAbsolutePath());
+      File starTreeV2File = new File(indexDir, V1Constants.STAR_TREE_V2_INDEX_FILE);
+      if (starTreeV2File.exists()) {
+        LOGGER.info("Replacing star tree v1 format with v2 format for {}", starTreeFile.getAbsolutePath());
+        FileUtils.copyFile(starTreeV2File, starTreeFile);
+      } else {
+        StarTreeInterf starTreeV1 = fromFile(starTreeFile, ReadMode.heap); // V1 only supports HEAP mode.
+
+        try {
+          writeTreeV2(starTreeV1, starTreeV2File);
+          FileUtils.copyFile(starTreeFile, new File(indexDir, V1Constants.STAR_TREE_V1_INDEX_FILE));
+          FileUtils.copyFile(starTreeV2File, starTreeFile);
+        } catch (Exception e) {
+          LOGGER.error("Exception caught while convert star tree v1 to v2 for {}", starTreeFile.getAbsolutePath(), e);
+        }
+      }
+    } else if (actualVersion == StarTreeFormatVersion.V2 && starTreeVersionToLoad == StarTreeFormatVersion.V1) {
+      File starTreeV1File = new File(indexDir, V1Constants.STAR_TREE_V1_INDEX_FILE);
+      if (starTreeV1File.exists()) {
+        try {
+          FileUtils.copyFile(starTreeFile, new File(indexDir, V1Constants.STAR_TREE_V2_INDEX_FILE));
+          FileUtils.copyFile(starTreeV1File, starTreeFile);
```

Comment:
This is only the conversion phase, and an exception here does not necessarily block loading. If something went really wrong such that the star tree binary was unloadable, then the loading phase (that happens after this convert phase) will throw an exception and the server will reload the segment from the controller.